crates/kubelet/src/node.rs

```diff
@@ -166,6 +167,124 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
     info!("Successfully created node '{}'", &config.node_name);
 }
 
+/// Fetch the uid of a node by name.
+pub async fn node_uid(client: &kube::Client, node_name: &str) -> anyhow::Result<String> {
+    let node_client: Api<KubeNode> = Api::all(client.clone());
+    match retry!(node_client.get(node_name).await, times: 4, log_error: |e| error!("Failed to get node to cordon: {:?}", e))
+    {
+        Ok(node) => match node.metadata.and_then(|meta| meta.uid) {
+            Some(uid) => Ok(uid),
+            None => {
+                let err = format!("Node missing metadata or uid {}.", node_name);
+                error!("{}", &err);
+                anyhow::bail!(err);
+            }
+        },
+        Err(e) => {
+            error!("Error fetching node {} id: {:?}.", node_name, e);
+            anyhow::bail!(e);
+        }
+    }
+}
+
+/// Cordons node and evicts all pods.
+pub async fn drain_node(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    cordon_node(client, node_name).await?;
+    evict_pods(client, node_name).await?;
+    Ok(())
+}
+
+/// Fetches list of pods on this node and deletes them.
+pub async fn evict_pods(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    let pod_client: Api<Pod> = Api::all(client.clone());
+    let node_selector = format!("spec.nodeName={}", node_name);
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    warn!("Evicting {} pods.", pods.len());
+
+    // TODO: Filter mirror pods,daemonsets, kube-system?
+    for pod in &pods {
+        if let Some(name) = pod.metadata.as_ref().and_then(|meta| meta.name.as_ref()) {
```

Comment:
You can bring the `Meta` trait into scope to avoid all of this unwrapping. Also, it may be worth it to convert all of these to our internal `Pod` wrapper to get convenient accessors on this data

Reply:
Yeah, I was on the fence about using the `Pod` wrapper, I will try it out.